File: app/src/main/java/io/treehouses/remote/Fragments/DialogFragments/WifiDialogFragment.java
package io.treehouses.remote.Fragments.DialogFragments;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.wifi.ScanResult;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import io.treehouses.remote.R;
import io.treehouses.remote.utils.ButtonConfiguration;

public class WifiDialogFragment extends DialogFragment {

    private AlertDialog mDialog;
    private WifiManager wifiManager;
    private ArrayList<String> wifiList = new ArrayList<>();
    private Context context;
    private String SSID;
    private View mView;
    private Boolean firstScan = true;
    private ProgressBar progressBar;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        LayoutInflater inflater = getActivity().getLayoutInflater();
        mView = inflater.inflate(R.layout.dialog_listview, null);
        progressBar = mView.findViewById(R.id.progressBar);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        context = getContext();

        setupWifi();

        mDialog = new AlertDialog.Builder(context)
                .setView(mView)
                .setIcon(R.drawable.dialog_icon)
                .setTitle("Choose a network: ")
                .setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                }).create();

        Log.e("TAG", "SSID = " +  wifiList.toString());

        return mDialog;
    }

    public static androidx.fragment.app.DialogFragment newInstance() {
        return new WifiDialogFragment();
    }

    private void setAdapter() {
        ListView listView = mView.findViewById(R.id.listView);
        ArrayAdapter<String> arrayAdapter = new ArrayAdapter<>(context, android.R.layout.simple_list_item_1, wifiList);

        listView.setAdapter(arrayAdapter);

        listView.setOnItemClickListener((parent, view, position, id) -> {
            SSID = wifiList.get(position);
            if (ButtonConfiguration.getEssid() == null) {
                ButtonConfiguration.getSSID().setText(SSID.trim());
            } else {
                ButtonConfiguration.getEssid().setText(SSID.trim());
            }
            wifiList.clear();
            dismiss();
        });
    }

    private void setupWifi() {
        wifiManager = (WifiManager) context.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        if (wifiManager == null) return;
        wifiManager.setWifiEnabled(true);
        BroadcastReceiver wifiScanReceiver = wifiBroadcastReceiver();

        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(WifiManager.SCAN_RESULTS_AVAILABLE_ACTION);
        context.registerReceiver(wifiScanReceiver, intentFilter);

        boolean success = wifiManager.startScan();
        if (!success) {
            scanFailure();
        }
    }

    private BroadcastReceiver wifiBroadcastReceiver() {
        return new BroadcastReceiver() {
            @Override
            public void onReceive(Context c, Intent intent) {
                boolean success = intent.getBooleanExtra(WifiManager.EXTRA_RESULTS_UPDATED, false);
                if (success) {
                    scanSuccess();
                } else if (firstScan) {
                    scanFailure();
                }
            }
        };
    }

    private void getSSIDs(List<ScanResult> results) {
        wifiList.clear();
        // converts Object list to array
        Object[] object = results.toArray();
        String temp = Arrays.toString(object);
        String[] resultArray = temp.split(",");

        // extracts SSID from wifi data
        for (String s : resultArray) {
            if (s.contains("SSID") && !s.contains("BSSID")) {

                String ssid = s.substring(6);

                // add to list if SSID is not hidden
                addToList(ssid);

                Log.e("TAG", "SSID = " + ssid);
            }
        }
    }

    private void addToList(String ssid) {
        if (ssid.trim().length()>0) {
            wifiList.add(ssid);
            progressBar.setVisibility(View.INVISIBLE);
        }
    }

    private void scanSuccess() {
        List<ScanResult> results = wifiManager.getScanResults();
        Log.e("TAG", "Scan Success - scan results: " + results);

        getSSIDs(results);
        setAdapter();
    }

    private void scanFailure() {
        // handle failure: new scan did not succeed
        List<ScanResult> results = wifiManager.getScanResults();
        Log.e("TAG", "Scan Failed - scan results: " + results);

        getSSIDs(results);

        if (results.size() >= 1 && firstScan) {
            Toast.makeText(context, "Scan unsuccessful. These are old results", Toast.LENGTH_LONG).show();
            setAdapter();
        } else if (results.size() < 1 && firstScan) {
            ifResultListEmpty();
        }
        firstScan = false;

    }

    private void ifResultListEmpty() {
        Toast.makeText(context, "Scan unsuccessful, please try again.", Toast.LENGTH_LONG).show();
        dismiss();
    }
}

File: app/src/main/kotlin/io/treehouses/remote/fragments/dialogfragments/WifiDialogFragment.kt
package io.treehouses.remote.fragments.dialogfragments

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.*
import android.net.wifi.ScanResult
import android.net.wifi.WifiManager
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import androidx.annotation.RequiresApi
import androidx.fragment.app.DialogFragment
import io.treehouses.remote.R
import io.treehouses.remote.utils.DialogUtils
import java.util.*


class WifiDialogFragment : DialogFragment() {
    private var mDialog: AlertDialog? = null
    private var wifiManager: WifiManager? = null
    private val wifiList = ArrayList<String>()
    private var mContext: Context? = null
    private var SSID: String? = null
    private var mView: View? = null
    private var firstScan = true
    private var progressBar: ProgressBar? = null
    override fun onAttach(context: Context) {
        super.onAttach(context)
        val inflater = requireActivity().layoutInflater
        mView = inflater.inflate(R.layout.dialog_listview, null)
        progressBar = mView!!.findViewById(R.id.progressBar)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        mContext = context
        setupWifi()
        mDialog = DialogUtils.createAlertDialog(context, mView, R.drawable.dialog_icon)
                .setTitle("Choose a network: ")
                .setNegativeButton("Cancel") { dialog: DialogInterface, _: Int -> dialog.dismiss() }.create()
        mDialog!!.window!!.setBackgroundDrawableResource(android.R.color.transparent)
        return mDialog!!
    }

    private fun setAdapter() {
        val listView = mView!!.findViewById<ListView>(R.id.listView)
        val arrayAdapter = ArrayAdapter(mContext!!, R.layout.simple_list_item, wifiList)
        listView.adapter = arrayAdapter
        listView.onItemClickListener = OnItemClickListener { _: AdapterView<*>?, _: View?, position: Int, _: Long ->
            SSID = wifiList[position]
            if (targetFragment != null) {
                val intent = Intent()
                intent.putExtra(WIFI_SSID_KEY, SSID!!.trim { it <= ' ' })
                targetFragment!!.onActivityResult(targetRequestCode, Activity.RESULT_OK, intent)
                wifiList.clear()
                dismiss()
            }
        }
    }

    private fun setupWifi() {
        wifiManager = mContext!!.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
        if (wifiManager == null) return
        wifiManager!!.isWifiEnabled = true
        val wifiScanReceiver = wifiBroadcastReceiver()
        val intentFilter = IntentFilter()
        intentFilter.addAction(WifiManager.SCAN_RESULTS_AVAILABLE_ACTION)
        mContext!!.registerReceiver(wifiScanReceiver, intentFilter)
        val success = wifiManager!!.startScan()
        if (!success) {
            scanFailure()
        }
    }

    private fun wifiBroadcastReceiver(): BroadcastReceiver {
        return object : BroadcastReceiver() {
            @RequiresApi(api = Build.VERSION_CODES.M)
            override fun onReceive(c: Context, intent: Intent) {
                val success = intent.getBooleanExtra(WifiManager.EXTRA_RESULTS_UPDATED, false)
                if (success) {
                    scanSuccess()
                } else if (firstScan) {
                    scanFailure()
                }
            }
        }
    }

    private fun getSSIDs(results: List<ScanResult>) {
        wifiList.clear()
        // converts Object list to array
        val `object`: Array<Any> = results.toTypedArray()
        val temp = `object`.contentToString()
        val resultArray = temp.split(",".toRegex()).toTypedArray()

        // extracts SSID from wifi data
        for (s in resultArray) {
            if (s.contains("SSID") && !s.contains("BSSID")) {
                val ssid = s.substring(6)

                // add to list if SSID is not hidden
                addToList(ssid)
            }
        }
    }

    private fun addToList(ssid: String) {
        if (ssid.trim { it <= ' ' }.isNotEmpty()) {
            wifiList.add(ssid)
            progressBar!!.visibility = View.INVISIBLE
        }
    }

    private fun scanSuccess() {
        val results = wifiManager!!.scanResults
        getSSIDs(results)
        setAdapter()
    }

    private fun scanFailure() {
        // handle failure: new scan did not succeed
        val results = wifiManager!!.scanResults
        getSSIDs(results)
        if (results.size >= 1 && firstScan) {
            Toast.makeText(context, "Scan unsuccessful. These are old results", Toast.LENGTH_LONG).show()
            setAdapter()
        } else if (results.size < 1 && firstScan) {
            ifResultListEmpty()
        }
        firstScan = false
    }

    private fun ifResultListEmpty() {
        Toast.makeText(context, "Scan unsuccessful, please try again.", Toast.LENGTH_LONG).show()
        dismiss()
    }

    companion object {
        var WIFI_SSID_KEY = "SSID"
        fun newInstance(): DialogFragment {
            return WifiDialogFragment()
        }
    }
}